### ISSUE TITLE
StackOverflowError with contentConverter that returns array/list type (same as #357)

### ISSUE THREAD

**Issue by Quurks** [FasterXML/jackson-databind#1704]
Java Version: 1.8
Affected Jackson versions: at least from 2.8.0 to 2.8.9, also 2.8.10-SNAPSHOT
**Unaffected** Jackson versions: 2.9.0.pr1 - 2.9.0.pr4 

This is exactly the same problem as #357, contentConverter creates a StackOverflowException, using the exact same code.

``` java
public class Test {
    //Example 1: List
    static class Example1 {
        @JsonSerialize(contentConverter = Converter1.class)
        public List<String> list = Collections.emptyList();
    }

    static class Converter1 extends StdConverter<String, List<Integer>> {
        @Override public List<Integer> convert(String value) {
            return Collections.emptyList();
        }
    }

    //Example 2: Array
    static class Example2 {
        @JsonSerialize(contentConverter = Converter2.class)
        public List<String> list = Collections.emptyList();
    }

    static class Converter2 extends StdConverter<String, String[]> {
        @Override public String[] convert(String value) {
            return new String[0];
        }
    }

    //Example 3: Exactly the code from #357
    static class A {}

    static class B {
        @JsonSerialize(contentConverter = AToStringConverter.class)
        public List<A> list = Arrays.asList(new A());
    }

    static class AToStringConverter extends StdConverter<A, List<String>> {
        @Override public List<String> convert(A value) {
            return Arrays.asList("Hello world.");
        }
    }

    public static void main(String[] args) throws JsonProcessingException {
        System.out.println(PackageVersion.VERSION);
        //All three examples throw the exception
        new ObjectMapper().writeValueAsString(new Example1());
        new ObjectMapper().writeValueAsString(new Example2());
        new ObjectMapper().writeValueAsString(new B());
    }
}
```

**Comment by cowtowncoder** [comment]
This is #357, which is only fixed for 2.9. It will not be fixed for earlier versions, due to concerns that fix might break other parts of code.

**Comment by cowtowncoder** [comment]
Actually forgot to ask: is this literally same thing that #357 had (and that was fixed); or possible follow-up (that was alluded to)? As per my first comment, part that was fixed will not be backported.
But if there is something different for 2.9.0 I would want to follow up.

**Comment by Quurks** [comment]
It's the exact same thing. 
Looking at the date of your comments on the old issue I was assuming the fix was in 2.8.X - my bad.

**Comment by cowtowncoder** [comment]
@Quurks ah. Makes sense, np.
